allennlp/models/semantic_role_labeller.py

```diff
@@ -0,0 +1,191 @@
+            # Negative log likelihood criterion takes integer labels, not one hot.
+            if tags.dim() == 3:
+                _, tags = tags.max(-1)
+            loss = self.sequence_loss(reshaped_log_probs, tags.view(-1))
+            output_dict["loss"] = loss
+
+        return output_dict
+    # pylint: enable=arguments-differ
+
+    def tag(self, text_field: TextField, verb_indicator: IndexField) -> Dict[str, Any]:
+        """
+        Perform inference on a ``Instance`` consisting of a single ``TextField`` representing
+        the sentence and an ``IndexField`` representing an optional index into the sentence
+        denoting a verbal predicate.
+
+        Returned sequence is the maximum likelihood tag sequence under the constraint that
+        the sequence must be a valid BIO sequence.
+
+        Parameters
+        ----------
+        text_field : ``TextField``, required.
+            A ``TextField`` containing the text to be tagged.
+        verb_indicator: ``IndexField``, required.
+            The index of the verb whose arguments we are labelling.
+
+        Returns
+        -------
+        A Dict containing:
+
+        tags : List[str]
+            A list the length of the text input, containing the predicted (argmax) tag
+            from the model per token.
+        class_probabilities : numpy.Array
+            An array of shape (text_input_length, num_classes), where each row is a
+            distribution over classes for a given token in the sentence.
+        """
+        instance = Instance({"tokens": text_field, "verb_indicator": verb_indicator})
+        instance.index_fields(self.vocab)
+        model_input = instance.as_array(instance.get_padding_lengths())
+        torch_input = arrays_to_variables(model_input)
+        # TODO(Mark): Make the data API always return tensors with batch dimensions at every abstraction level.
+        # Add a batch dimension by unsqueezing, because pytorch doesn't support inputs without one.
+        torch_input["tokens"]["tokens"].data.unsqueeze_(0)
+        torch_input["verb_indicator"].data.unsqueeze_(0)
+
+        output_dict = self.forward(**torch_input)
+
+        # Remove batch dimension, as we only had one input.
+        predictions = output_dict["class_probabilities"].data.squeeze(0)
+        transition_matrix = self.get_viterbi_pairwise_potentials()
+
+        max_likelihood_sequence, _ = viterbi_decode(predictions, transition_matrix)
+        tags = [self.vocab.get_token_from_index(x, namespace="tags")
+                for x in max_likelihood_sequence]
+
+        return {"tags": tags, "class_probabilities": predictions.numpy()}
+
+    def get_viterbi_pairwise_potentials(self):
+        """
+        Generate a matrix of pairwise transition potentials for the BIO labels.
+        The only constraint implemented here is that I-XXX tags must be preceded
+        by either an identical I-XXX tag or a B-XXX tag (i.e they have a pairwise
+        potential of -inf).
```

Comment:
Yeah, it's just the antecedent of "they" that's unclear.  It looks like it's referring to the `I-XXX` tag and a correct tag preceding it, which would have a potential of 0, while the potential of `-inf` applies to the `I-XXX` tag and _impossible_ preceding tags.